src/solver/td_simplified_ref.ml

```diff
@@ -0,0 +1,220 @@
+(** Top-down solver with side effects. Baseline for comparisons with td_parallel solvers     ([td_simplified_ref]).
+    This is the same as ([td_simplified]), but it uses records for solver that instead of multiple hashmaps.
+*)
+
+open Batteries
+open Goblint_constraint.ConstrSys
+open Goblint_constraint.SolverTypes
+open Messages
+
+module M = Messages
+
+module Base : GenericEqSolver =
+  functor (S:EqConstrSys) ->
+  functor (HM:Hashtbl.S with type key = S.v) ->
+  struct
+    open SolverBox.Warrow (S.Dom)
+    include Generic.SolverStats (S) (HM)
+    module VS = Set.Make (S.Var)
+
+    type var_data = {
+      infl: VS.t;
+      value: S.Dom.t;
+      wpoint: bool;
+      stable: bool;
+      called: bool;
+    }
+
+    let solve st vs =
+      let (data : var_data ref HM.t) = HM.create 10 in
+
+      let () = print_solver_stats := fun () ->
+          Logs.debug "|data|=%d" (HM.length data);
+      in
+
+      let add_infl y x =
+        if tracing then trace "infl" "add_infl %a %a" S.Var.pretty_trace y S.Var.pretty_trace x;
+        let y_ref = HM.find data y in
+        y_ref := { !y_ref with infl = VS.add x !y_ref.infl }
+      in
+
+      let init (x : S.v): var_data ref =
+        let x_ref = HM.find_option data x in
+        match x_ref with
+        | Some r -> r
+        | None ->
+          begin
+            new_var_event x;
+            if tracing then trace "init" "init %a" S.Var.pretty_trace x;
+            let data_x = ref { 
+                infl = VS.empty;
+                value = S.Dom.bot ();
+                wpoint = false;
+                stable = false;
+                called = false 
+              } in
+            HM.replace data x data_x;
+            data_x
+          end
```

Comment:
`BatOption.default_delayed` ?

Reply:
(as discussed offline) It would do the same, but to me it implies we are primarily returning the value, whereas the `HM.replace ...` is significant here, so I prefer this version.